# Part 1. Equation of a Slime

How many late days are you using for this assignment? 
0

In [1]:
# Imports section
import pandas as pd              
import numpy as np                
import sklearn as sk

## 1. Loading the dataset

In [2]:
# Using pandas load the dataset
df = pd.read_csv('science_data_large.csv')  

# Output the first 15 rows of the data
print(df.head(15))

# Display a summary of the table information (number of datapoints, etc.)
print(df.info())
print(df.describe())

    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 n

## 2. Splitting the dataset

In [3]:
# Take the pandas dataset and split it into our features (X) and label (y)
X = df.drop('Size nm^3', axis=1)
y = df['Size nm^3']


# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
# For grading consistency use random_state=42 

X_train, X_test, y_train, y_test = sk.model_selection.train_test_split(X, y, test_size=0.1, random_state=42)

## 3. Perform a Linear Regression

In [4]:
# Use sklearn to train a model on the training set

# Create a sample datapoint and predict the output of that sample with the trained model

# Report the score for that model using the default score function property of the SKLearn model, in your own words (markdown, not code) explain what the score means

# Extract the coefficients and intercept from the model and write an equation for your h(x) using LaTeX

model = sk.linear_model.LinearRegression()
model.fit(X_train, y_train)

sample = pd.DataFrame([[0.5, 150]], columns=X_train.columns)    # Replace with appropriate sample values
prediction = model.predict(sample)
print("Sample Prediction:", prediction)

score = model.score(X_test, y_test)
print("Model Score:", score)

coefficients = model.coef_
intercept = model.intercept_

print("Coefficients:", coefficients)
print("Intercept:", intercept)


Sample Prediction: [-254054.14640385]
Model Score: 0.8552472077276095
Coefficients: [ 866.14641337 1032.69506649]
Intercept: -409391.47958340764


Write the linear equation of a slime: (example equation: $E = mc^2$)



h(x) = 866.15⋅x1 + 1032.70⋅x2 − 409391.48


Report on score and explain meaning:

The model achieved an R^2 score of 0.8552, which indicates that it explains approximately 85.5% of the variance in the test data. This suggests the model has a good fit, meaning it captures most of the relationship between the features and the target variable. However, it doesn't perfectly describe the data, leaving about 14.5% of the variance unexplained, which could be due to noise, missing features, or model limitations.

## 4. Use Cross Validation

In [5]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
# For grading consistency use n_splits=5 and random_state=42
scores = sk.model_selection.cross_val_score(model, X, y, cv=sk.model_selection.KFold(n_splits=5, random_state=42, shuffle=True))
print("Cross-validation scores:", scores)
print("Average score:", np.mean(scores))
print("Standard deviation:", np.std(scores))

# Report on their finding and their significance

Cross-validation scores: [0.86151889 0.82742341 0.87195173 0.88166206 0.85609101]
Average score: 0.8597294202684646
Standard deviation: 0.01838773713930639


Write findings here:
The cross-validation results show that the model achieved consistently high performance across the five folds, with R^2
  scores of 0.8615, 0.8274, 0.8720, 0.8817, and 0.8561. The average score of 0.8597 indicates that the model explains approximately 86% of the variance in the slime growth data. The low standard deviation of 0.0184 demonstrates that the model's performance is stable and reliable across different data splits. This consistency suggests that the model generalizes well and is unlikely to overfit, making it a strong predictor for estimating the slime's growth based on KCl concentration and temperature.


## 5. Using Polynomial Regression

In [6]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
# Perform k-fold cross validation (as above)

# Report on the metrics and output the resultant equation as you did in Part 3.
poly = sk.preprocessing.PolynomialFeatures(degree=2)
X_poly = poly.fit_transform(X)

X_train_poly, X_test_poly, y_train_poly, y_test_poly = sk.model_selection.train_test_split(
    X_poly, y, test_size=0.1, random_state=42
)

model_poly = sk.linear_model.LinearRegression()
model_poly.fit(X_train_poly, y_train_poly)

scores_poly = sk.model_selection.cross_val_score(
    model_poly, X_poly, y, cv=sk.model_selection.KFold(n_splits=5, random_state=42, shuffle=True)
)

print("Cross-validation scores:", scores_poly)
print("Average score:", np.mean(scores_poly))
print("Standard deviation:", np.std(scores_poly))
print("Intercept:", model_poly.intercept_)
print("Coefficients:", model_poly.coef_)


Cross-validation scores: [1. 1. 1. 1. 1.]
Average score: 1.0
Standard deviation: 0.0
Intercept: 2.0478793885558844e-05
Coefficients: [ 0.00000000e+00  1.20000000e+01 -1.27196606e-07  1.26450728e-11
  2.00000000e+00  2.85714287e-02]


Write the polynomial equation of a slime: (example equation: $E = mc^2$)


h(x) = 12.00000⋅x₁ + 2.00000⋅x₁⋅x₂ + 0.02857⋅x₂²

Report on the score and interpret:


The model achieved perfect cross-validation scores:
- Cross-validation scores: [1.0, 1.0, 1.0, 1.0, 1.0]
- Average score: 1.0
- Standard deviation: 0.0

An \(R^2\) score of 1.0 across all folds means that the model perfectly explains the variance in the data. In other words, the polynomial equation captures the relationship between the features and the target variable with no error. While this is an ideal result, it is important to consider that such perfect scores may also indicate a small or noise-free dataset or potential overfitting. If the model were to be applied to new data, overfitting could result in less accurate predictions.

# Part 2. Chronic Kidney Disease Prediction via Classification

Create code and markdown cells as needed to perform classification and report on your results

In [7]:
# Load the dataset. Then train and evaluate the classification models.

df_ckd = pd.read_csv('ckd_feature_subset.csv')
print(df_ckd.head())


X_ckd = df_ckd.drop('Target_ckd', axis=1)
y_ckd = df_ckd['Target_ckd']

# Define the classification models using the sk alias
models = {
    "Logistic Regression": sk.linear_model.LogisticRegression(random_state=42, max_iter=1000),
    "Support Vector Machine": sk.svm.SVC(random_state=42),
    "k-Nearest Neighbors": sk.neighbors.KNeighborsClassifier(),
    "Neural Network": sk.neural_network.MLPClassifier(random_state=42, max_iter=1000)
}

# Set up 5-fold cross-validation with a fixed random seed
kf = sk.model_selection.KFold(n_splits=5, random_state=42, shuffle=True)

# Evaluate each model and collect cross-validation metrics
results = {}
for name, model in models.items():
    scores = sk.model_selection.cross_val_score(model, X_ckd, y_ckd, cv=kf)
    results[name] = (np.mean(scores), np.std(scores))
    print(f"{name}: Mean Accuracy = {np.mean(scores):.4f}, Std Deviation = {np.std(scores):.4f}")

        age        bp      wbcc  appet_poor  appet_good      rbcc  Target_ckd
0  0.688312  0.333333  0.000000           1           0  0.000000           1
1  0.545455  0.333333  0.128319           1           0  0.305085           1
2  0.714286  0.500000  0.238938           1           0  0.186441           1
3  0.688312  0.333333  0.283186           0           1  0.338983           1
4  0.441558  0.333333  0.221239           1           0  0.220339           1
Logistic Regression: Mean Accuracy = 0.8566, Std Deviation = 0.0663
Support Vector Machine: Mean Accuracy = 0.9282, Std Deviation = 0.0476
k-Nearest Neighbors: Mean Accuracy = 0.9280, Std Deviation = 0.0524
Neural Network: Mean Accuracy = 0.9351, Std Deviation = 0.0405


## Results and Conclusion for Classification Experiments

### Classification Model Performance on the CKD Dataset

The following table summarizes the cross-validation results for the four classification models evaluated on the CKD dataset:

| Model                   | Mean Accuracy | Standard Deviation |
|-------------------------|---------------|--------------------|
| Logistic Regression     | 0.8566        | 0.0663             |
| Support Vector Machine  | 0.9282        | 0.0476             |
| k-Nearest Neighbors     | 0.9280        | 0.0524             |
| Neural Network          | 0.9351        | 0.0405             |

**Interpretation:**

- The **Neural Network** achieved the highest mean accuracy (0.9351) with the lowest standard deviation (0.0405), indicating not only superior predictive performance but also very consistent results across the folds.
- Both the **Support Vector Machine** and **k-Nearest Neighbors** models performed well, with mean accuracies of around 0.928 and moderate consistency.
- **Logistic Regression** showed the lowest performance among the four, with a mean accuracy of 0.8566 and a slightly higher standard deviation.

Overall, the neural network appears to be the best performing model on this CKD dataset, suggesting that its ability to capture non-linear patterns may be advantageous in predicting CKD status.


In [8]:
# Experiments with Neural Network.
nn_configs = {
    "NN Config 1 (50 neurons)": sk.neural_network.MLPClassifier(hidden_layer_sizes=(50,), random_state=42, max_iter=1000),
    "NN Config 2 (100 neurons)": sk.neural_network.MLPClassifier(hidden_layer_sizes=(100,), random_state=42, max_iter=1000),
    "NN Config 3 (50, 50 neurons)": sk.neural_network.MLPClassifier(hidden_layer_sizes=(50,50), random_state=42, max_iter=2000)
}

nn_results = {}
for name, model in nn_configs.items():
    scores = sk.model_selection.cross_val_score(model, X_ckd, y_ckd, cv=kf)
    nn_results[name] = (np.mean(scores), np.std(scores))
    print(f"{name}: Mean Accuracy = {np.mean(scores):.4f}, Std Deviation = {np.std(scores):.4f}")


NN Config 1 (50 neurons): Mean Accuracy = 0.9351, Std Deviation = 0.0405
NN Config 2 (100 neurons): Mean Accuracy = 0.9351, Std Deviation = 0.0405
NN Config 3 (50, 50 neurons): Mean Accuracy = 0.9546, Std Deviation = 0.0437


## Results and Conclusion for Neural Network Experiments

### Neural Network Configuration Performance

| Configuration               | Mean Accuracy | Standard Deviation |
|-----------------------------|---------------|--------------------|
| NN Config 1 (50 neurons)    | 0.9351        | 0.0405             |
| NN Config 2 (100 neurons)   | 0.9351        | 0.0405             |
| NN Config 3 (50, 50 neurons) | 0.9546        | 0.0437             |

**Interpretation:**

- **NN Config 1 (50 neurons)** and **NN Config 2 (100 neurons)** achieved identical performance, indicating that simply increasing the number of neurons in a single hidden layer from 50 to 100 did not improve classification accuracy for this dataset.
- **NN Config 3 (50, 50 neurons)** outperformed the single-layer configurations, reaching a higher mean accuracy (0.9546). This suggests that a deeper architecture with two hidden layers may better capture the underlying patterns in the CKD data.

